File: myfschool_mobile/lib/models/attendance.dart
class AttendanceRecord {
  final int? id;
  final int? studentId;
  final String studentName;
  final String studentCode;
  final int? classId;
  final String className;
  final int? subjectId;
  final String subjectName;
  final DateTime attendanceDate;
  final int? slotNumber;
  final String status; // 'PRESENT', 'EXCUSED_ABSENCE', 'UNEXCUSED_ABSENCE', 'LATE'
  final String note;
  final String recordedByName;

  AttendanceRecord({
    this.id,
    this.studentId,
    required this.studentName,
    required this.studentCode,
    this.classId,
    required this.className,
    this.subjectId,
    required this.subjectName,
    required this.attendanceDate,
    this.slotNumber,
    required this.status,
    required this.note,
    required this.recordedByName,
  });

  factory AttendanceRecord.fromJson(Map<String, dynamic> json) {
    return AttendanceRecord(
      id: json['id'] as int?,
      studentId: json['studentId'] as int?,
      studentName: json['studentName'] as String? ?? '',
      studentCode: json['studentCode'] as String? ?? '',
      classId: json['classId'] as int?,
      className: json['className'] as String? ?? '',
      subjectId: json['subjectId'] as int?,
      subjectName: json['subjectName'] as String? ?? '',
      attendanceDate: json['attendanceDate'] != null
          ? DateTime.parse(json['attendanceDate'].toString())
          : DateTime.now(),
      slotNumber: json['slotNumber'] as int?,
      status: json['status'] as String? ?? 'PRESENT',
      note: json['note'] as String? ?? '',
      recordedByName: json['recordedByName'] as String? ?? '',
    );
  }

  Map<String, dynamic> toJson() {
    return {
      if (id != null) 'id': id,
      if (studentId != null) 'studentId': studentId,
      'studentName': studentName,
      'studentCode': studentCode,
      if (classId != null) 'classId': classId,
      'className': className,
      if (subjectId != null) 'subjectId': subjectId,
      'subjectName': subjectName,
      'attendanceDate': attendanceDate.toIso8601String().split('T')[0],
      if (slotNumber != null) 'slotNumber': slotNumber,
      'status': status,
      'note': note,
      'recordedByName': recordedByName,
    };
  }

  String get statusLabel {
    switch (status.toUpperCase()) {
      case 'PRESENT':
        return 'Có mặt';
      case 'EXCUSED_ABSENCE':
        return 'Nghỉ có phép';
      case 'UNEXCUSED_ABSENCE':
        return 'Nghỉ không phép';
      case 'LATE':
        return 'Đi muộn';
      default:
        return status;
    }
  }
}

File: myfschool_frontend/src/types/attendance.ts
export type AttendanceStatus = 'PRESENT' | 'EXCUSED_ABSENCE' | 'UNEXCUSED_ABSENCE' | 'LATE';

export interface AttendanceRecordDTO {
  id: number;
  studentId: number;
  studentName: string;
  studentCode: string;
  classId?: number;
  className?: string;
  subjectId?: number;
  subjectName?: string;
  attendanceDate: string; // LocalDate format: "YYYY-MM-DD"
  slotNumber?: number;
  status: AttendanceStatus;
  note?: string;
  recordedByName?: string;
}

export interface AttendanceSessionSummaryDTO {
  attendanceDate: string; // LocalDate format: "YYYY-MM-DD"
  slotNumber: number;
  subjectId?: number;
  subjectName?: string;
  presentCount: number;
  excusedCount: number;
  unexcusedCount: number;
  lateCount: number;
  totalCount: number;
  canEdit?: boolean;
  absentStudents?: string[];
}

export interface AttendanceStudentSummaryDTO {
  studentId: number;
  studentName: string;
  studentCode: string;
  presentCount: number;
  excusedCount: number;
  unexcusedCount: number;
  lateCount: number;
  totalTrackedSessions: number;
  attendanceRate: number;
  isAtRisk?: boolean;
  warningNote?: string;
  records?: AttendanceRecordDTO[];
}

export interface AttendanceClassHistoryDTO {
  classId: number;
  className: string;
  totalStudents: number;
  attendanceRate: number;
  totalSessions: number;
  presentCount: number;
  excusedCount: number;
  unexcusedCount: number;
  lateCount: number;
  sessions: AttendanceSessionSummaryDTO[];
  atRiskStudents?: Array<{
    studentId: number;
    studentName: string;
    studentCode: string;
    unexcusedCount: number;
    rate: number;
    warningNote: string;
  }>;
  studentSummaries: AttendanceStudentSummaryDTO[];
}

export interface AttendanceSheetStudentDTO {
  studentId: number;
  studentName: string;
  studentCode: string;
  currentStatus?: AttendanceStatus;
  hasApprovedLeave?: boolean;
  leaveRequestId?: number;
  leaveReason?: string;
  note?: string;
}

export interface AttendanceSheetDTO {
  classId: number;
  className: string;
  subjectId?: number;
  subjectName?: string;
  slotNumber: number;
  startTime?: string; // LocalTime format: "HH:mm"
  endTime?: string;   // LocalTime format: "HH:mm"
  attendanceDate: string; // LocalDate format: "YYYY-MM-DD"
  canEdit: boolean;
  lockReason?: string; // "NOT_STARTED" | "EXPIRED_PAST_DAY" | "NO_SCHEDULE" | "FUTURE_DATE"
  totalStudents: number;
  students: AttendanceSheetStudentDTO[];
}

export interface AttendanceBatchItemDTO {
  studentId: number;
  status: AttendanceStatus;
  note?: string;
  overrideLeave?: boolean;
  overrideReason?: string;
}

export interface AttendanceBatchRequest {
  classId: number;
  subjectId?: number;
  slotNumber: number;
  attendanceDate: string; // LocalDate format: "YYYY-MM-DD"
  items: AttendanceBatchItemDTO[];
}

export interface UnrecordedAttendanceSessionDTO {
  scheduleId: number;
  classId: number;
  className: string;
  subjectId?: number;
  subjectName?: string;
  teacherId?: number;
  teacherName?: string;
  slotNumber: number;
  startTime?: string;
  endTime?: string;
  totalStudents: number;
  recordedStudents: number;
  delayMinutes: number;
  delayFormatted: string;
  attendanceStatus: 'NOT_ATTENDED' | 'PARTIALLY_ATTENDED';
  alertStatus: 'CHUA_TRE' | 'CANH_BAO_TRE' | 'QUA_HAN';
  alertStatusLabel: string;
  reminderSent: boolean;
}
